brewblox_devcon_spark/commands.py

```diff
@@ -114,247 +98,311 @@ class Command(ABC):
     The constructor will fail if these are not defined.
 
     Required class variables:
-        _OPCODE
-        _REQUEST
-        _RESPONSE
+        _OPCODE: OpcodeEnum
+        _REQUEST: Struct
+        _RESPONSE: Struct
+        _VALUES: Tuple[str, Struct]
+
+    _OPCODE must always be set, _REQUEST, _RESPONSE, and _VALUES can be None.
+
+    _REQUEST or _RESPONSE being None does not mean the controller will literally send or receive nothing.
+    opcode is always sent, and errcode is always received.
+
+    _VALUES must be formatted as Tuple[str, Struct]. The string indicates the key in the decoded response.
+
+    Example:
 
-    Opcode must always be set, request and response can be None.
+        class ExampleCommand(Command):
+            _OPCODE = OpcodeEnum.UNUSED
+            _REQUEST = Struct('first_arg' / Int8sb) + Struct('second_arg' / Int8sb)
+            _RESPONSE = None
+            _VALUES = ('response_list', Int8sb)
 
-    Request or response being None does not mean the controller will literally send nothing.
-    Opcode/error code are always sent, and will be part of the decoded request/response.
+        decoded = {
+            'first_arg': 1,
+            'second_arg': 2,
+            'response_list': [4, 5, 6]
+        }
     """
 
-    def __init__(self):
-        opcode = self.__class__._OPCODE
-        self._opcode = Struct('opcode' / Const(OpcodeEnum.encmapping[opcode], Byte))
+    def __init__(self, encoded=(None, None), decoded=(None, None)):
+        self._encoded_request, self._encoded_response = encoded
+        self._decoded_request, self._decoded_response = decoded
+
+        self._check_sanity()
+
+        # `opcode` is defined as a private variable in request.
+        # it will be included in `self.encoded_request`, but not in `self.decoded_request`.
+        # `opcode` values are linked to the class, and will never deviate from the class-defined value.
+        opcode_val = self.__class__._OPCODE
+        opcode = Struct('_opcode' / Const(OpcodeEnum.encmapping[opcode_val], Byte))
+
+        # `errcode` is defined as a private variable in response.
```

Comment:
comment that errors are handled as exceptions

Reply:
fixed